// @webex/react-component-join-call-button/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Button from '@webex/react-component-button';
import Timer from '@webex/react-component-timer';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-join-call-button__join__LmpvaW4g {\n  width: 55px;\n  height: 47.5px;\n}\n\n.react-component-join-call-button__joinButton__LmpvaW4g {\n  height: 25px;\n  width: 55px;\n  background-color: #30d557;\n  color: #fff;\n  border-radius: 25px;\n}\n\n.react-component-join-call-button__duration__LmpvaW4g {\n  color: #30d557;\n  text-align: center;\n  font-size: 13px;\n}\n";
var styles = {"join":"react-component-join-call-button__join__LmpvaW4g","joinButton":"react-component-join-call-button__joinButton__LmpvaW4g","duration":"react-component-join-call-button__duration__LmpvaW4g"};
styleInject(css_248z);

function JoinCallButton({
  callStartTime,
  onJoinClick
}) {
  return /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-join-call', styles.join)
  }, /*#__PURE__*/React.createElement(Button, {
    accessibilityLabel: "Join Call",
    buttonClassName: classNames('webex-join-call-button', styles.joinButton),
    label: "Join",
    onClick: onJoinClick
  }), /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-join-call-duration', styles.duration)
  }, /*#__PURE__*/React.createElement(Timer, {
    startTime: callStartTime
  })));
}

JoinCallButton.propTypes = {
  callStartTime: PropTypes.number.isRequired,
  onJoinClick: PropTypes.func.isRequired
};

export default JoinCallButton;
