// @webex/react-component-title-bar/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import PresenceAvatar from '@webex/react-container-presence-avatar';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-title-bar__titleBar__LnRpdGxl {\n  display: flex;\n  height: 46px;\n  border-bottom: 1px solid #ececed;\n}\n\n.react-component-title-bar__title__LnRpdGxl {\n  margin: 0;\n  overflow: hidden;\n  font-size: 12px;\n  line-height: 16px;\n  text-overflow: ellipsis;\n  white-space: nowrap;\n  background-color: transparent;\n  flex: 1 1 auto;\n}\n\n.react-component-title-bar__avatarContainer__LnRpdGxl {\n  display: flex;\n  width: 24px;\n  height: 24px;\n  margin: 11px 12px 11px 14px;\n  flex-grow: 0;\n  flex-shrink: 0;\n}\n\n.react-component-title-bar__titleText__LnRpdGxl {\n  display: flex;\n  overflow: hidden;\n  align-items: center;\n  flex-grow: 2;\n}\n\n.react-component-title-bar__menuContainer__LnRpdGxl {\n  display: flex;\n  width: 40px;\n  padding-right: 18px;\n  align-items: center;\n  justify-content: center;\n}\n\n.react-component-title-bar__buttonContainer__LnRpdGxl {\n  width: 40px;\n  height: 40px;\n  border: 1px solid rgba(0, 0, 0, 0.06);\n  border-radius: 50%;\n}\n\n.react-component-title-bar__buttonContainer__LnRpdGxl:hover {\n  color: #fff;\n  background-color: #049fd9;\n}\n\n.react-component-title-bar__buttonContainer__LnRpdGxl button {\n  width: 100%;\n  height: 100%;\n  padding: 7px;\n  font-size: 24px;\n}\n\n.react-component-title-bar__buttonContainer__LnRpdGxl button:focus {\n  outline: none;\n}\n\n.react-component-title-bar__titleMeta__LnRpdGxl {\n  display: flex;\n  align-items: center;\n}\n";
var styles = {"titleBar":"react-component-title-bar__titleBar__LnRpdGxl","title":"react-component-title-bar__title__LnRpdGxl","avatarContainer":"react-component-title-bar__avatarContainer__LnRpdGxl","titleText":"react-component-title-bar__titleText__LnRpdGxl","menuContainer":"react-component-title-bar__menuContainer__LnRpdGxl","buttonContainer":"react-component-title-bar__buttonContainer__LnRpdGxl","titleMeta":"react-component-title-bar__titleMeta__LnRpdGxl"};
styleInject(css_248z);

const propTypes = {
  children: PropTypes.node,
  avatarId: PropTypes.string,
  name: PropTypes.string,
  type: PropTypes.string
};
const defaultProps = {
  children: undefined,
  avatarId: '',
  name: '',
  type: 'default'
};

function TitleBar({
  avatarId,
  children,
  name,
  type
}) {
  return /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-title-bar', styles.titleBar)
  }, /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-avatar-container', styles.avatarContainer)
  }, /*#__PURE__*/React.createElement(PresenceAvatar, {
    avatarId: avatarId,
    name: name,
    spaceType: type,
    size: 24
  })), /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-title-text', styles.titleText)
  }, /*#__PURE__*/React.createElement("p", null, /*#__PURE__*/React.createElement("strong", {
    className: classNames('webex-title', styles.title)
  }, name))), /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-title-meta', styles.titleMeta)
  }, children));
}

TitleBar.propTypes = propTypes;
TitleBar.defaultProps = defaultProps;

export default TitleBar;
