// @webex/react-component-video/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-video__video__LnZpZGVv {\n  width: 100%;\n  height: 100%;\n}\n";
var styles = {"video":"react-component-video__video__LnZpZGVv"};
styleInject(css_248z);

const propTypes = {
  audioMuted: PropTypes.bool,
  srcObject: PropTypes.object.isRequired
};
const defaultProps = {
  audioMuted: false
};

function Video({
  srcObject,
  audioMuted
}) {
  // Need autoPlay to start the video automagically
  if (!srcObject) {
    return null;
  }

  function getEl(el) {
    const element = el;

    if (element && srcObject) {
      element.srcObject = srcObject;
    }

    return element;
  }

  return (
    /*#__PURE__*/
    // eslint-disable-reason cannot provide caption for video stream yet
    // eslint-disable-next-line jsx-a11y/media-has-caption
    React.createElement("video", {
      autoPlay: true,
      playsInline: true,
      className: classNames(styles.video),
      muted: audioMuted,
      ref: getEl
    })
  );
}

Video.propTypes = propTypes;
Video.defaultProps = defaultProps;

export default Video;
