// @webex/react-component-add-file-button/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Icon, { ICONS } from '@webex/react-component-icon';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-add-file-button__container__LmNvbnRh {\n  position: relative;\n  width: 32px;\n  height: 32px;\n}\n\n.react-component-add-file-button__button__LmNvbnRh {\n  position: absolute;\n  width: 32px;\n  height: 32px;\n  text-align: center;\n  cursor: pointer;\n  background-color: #fff;\n  border: 1px solid rgba(0, 0, 0, 0.06);\n  border-radius: 50%;\n  outline: none;\n}\n\n.react-component-add-file-button__icon__LmNvbnRh {\n  width: 16px;\n  height: 16px;\n  font-size: 17px;\n  line-height: 14px;\n  color: #4f5051;\n}\n\n.react-component-add-file-button__fileInput__LmNvbnRh {\n  position: absolute;\n  width: 32px;\n  height: 32px;\n  overflow: hidden;\n  opacity: 0;\n}\n";
var styles = {"container":"react-component-add-file-button__container__LmNvbnRh","button":"react-component-add-file-button__button__LmNvbnRh","icon":"react-component-add-file-button__icon__LmNvbnRh","fileInput":"react-component-add-file-button__fileInput__LmNvbnRh"};
styleInject(css_248z);

const propTypes = {
  onChange: PropTypes.func
};
const defaultProps = {
  onChange: () => {}
};

function AddFileButton(props) {
  const {
    onChange
  } = props;
  return /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-add-file-container', styles.container)
  }, /*#__PURE__*/React.createElement("button", {
    className: classNames('webex-add-file-button', styles.button)
  }, /*#__PURE__*/React.createElement("span", {
    className: styles.icon
  }, /*#__PURE__*/React.createElement(Icon, {
    type: ICONS.ICON_TYPE_ADD
  }))), /*#__PURE__*/React.createElement("input", {
    className: classNames('webex-file-input', styles.fileInput),
    multiple: "multiple",
    onChange: onChange,
    type: "file"
  }));
}

AddFileButton.propTypes = propTypes;
AddFileButton.defaultProps = defaultProps;

export default AddFileButton;
