// @webex/react-component-incoming-call/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import PresenceAvatar from '@webex/react-container-presence-avatar';
import ButtonControls from '@webex/react-component-button-controls';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-incoming-call__callInactiveContainer__LmNhbGxJ {\n  position: relative;\n  display: flex;\n  margin: auto;\n  flex: 1 1 auto;\n  flex-direction: column;\n  align-items: center;\n}\n\n.react-component-incoming-call__personName__LmNhbGxJ {\n  font-size: 32px;\n  font-weight: 200;\n  color: #4f5051;\n}\n\n.react-component-incoming-call__callControls__LmNhbGxJ {\n  z-index: 100;\n  margin: 34px 16px 0;\n  text-align: center;\n}\n\n.react-component-incoming-call__incomingCallLabel__LmNhbGxJ {\n  font-size: 14px;\n  font-weight: 300;\n  color: #858688;\n}\n\n.react-component-incoming-call__callControls__LmNhbGxJ .react-component-incoming-call__answerButton__LmNhbGxJ {\n  background-color: #30d557;\n}\n\n.react-component-incoming-call__callControls__LmNhbGxJ .react-component-incoming-call__declineButton__LmNhbGxJ {\n  background-color: #ff513d;\n}\n";
var styles = {"callInactiveContainer":"react-component-incoming-call__callInactiveContainer__LmNhbGxJ","personName":"react-component-incoming-call__personName__LmNhbGxJ","callControls":"react-component-incoming-call__callControls__LmNhbGxJ","incomingCallLabel":"react-component-incoming-call__incomingCallLabel__LmNhbGxJ","answerButton":"react-component-incoming-call__answerButton__LmNhbGxJ","declineButton":"react-component-incoming-call__declineButton__LmNhbGxJ"};
styleInject(css_248z);

const propTypes = {
  answerButtonLabel: PropTypes.string,
  avatarId: PropTypes.string,
  avatarImage: PropTypes.string,
  declineButtonLabel: PropTypes.string,
  displayName: PropTypes.string.isRequired,
  incomingCallMessage: PropTypes.string,
  onAnswerClick: PropTypes.func.isRequired,
  onDeclineClick: PropTypes.func.isRequired
};
const defaultProps = {
  answerButtonLabel: '',
  avatarId: '',
  avatarImage: '',
  declineButtonLabel: '',
  incomingCallMessage: ''
};

function IncomingCall({
  answerButtonLabel,
  avatarId,
  avatarImage,
  declineButtonLabel,
  incomingCallMessage,
  onAnswerClick,
  onDeclineClick,
  displayName
}) {
  const buttons = [{
    accessibilityLabel: answerButtonLabel,
    label: answerButtonLabel,
    buttonType: 'camera',
    callControl: true,
    color: 'green',
    onClick: onAnswerClick
  }, {
    accessibilityLabel: declineButtonLabel,
    label: declineButtonLabel,
    buttonType: 'cancel',
    callControl: true,
    onClick: onDeclineClick
  }];
  return /*#__PURE__*/React.createElement("div", {
    className: classNames(styles.callInactiveContainer, 'call-inactive-container')
  }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(PresenceAvatar, {
    avatarId: avatarId,
    image: avatarImage,
    name: displayName,
    size: "40"
  })), /*#__PURE__*/React.createElement("div", {
    className: classNames(styles.personName, 'call-person-name')
  }, displayName), /*#__PURE__*/React.createElement("div", {
    className: classNames(styles.incomingCallLabel, 'incoming-call-label')
  }, incomingCallMessage), /*#__PURE__*/React.createElement("div", {
    className: classNames(styles.callControls, 'call-controls-container')
  }, /*#__PURE__*/React.createElement(ButtonControls, {
    buttons: buttons,
    showLabels: true
  })));
}

IncomingCall.propTypes = propTypes;
IncomingCall.defaultProps = defaultProps;

export default IncomingCall;
