// @webex/redux-module-indicators/es/index.js
import { fromJS } from 'immutable';

const ADD_TYPING_INDICATOR = 'indicators/ADD_TYPING_INDICATOR';
const DELETE_TYPING_INDICATOR = 'indicators/DELETE_TYPING_INDICATOR';
const initialState = fromJS({
  typing: {}
});
function reducer(state = initialState, action) {
  switch (action.type) {
    case ADD_TYPING_INDICATOR:
      return state.setIn(['typing', action.payload.userId], true);

    case DELETE_TYPING_INDICATOR:
      return state.deleteIn(['typing', action.payload.userId]);

    default:
      return state;
  }
}
function addTyping(userId) {
  return {
    type: ADD_TYPING_INDICATOR,
    payload: {
      userId
    }
  };
}
function deleteTyping(userId) {
  return {
    type: DELETE_TYPING_INDICATOR,
    payload: {
      userId
    }
  };
}
function setTyping(userId, value) {
  if (value) {
    return addTyping(userId);
  }

  return deleteTyping(userId);
}

export default reducer;
export { ADD_TYPING_INDICATOR, DELETE_TYPING_INDICATOR, addTyping, deleteTyping, initialState, setTyping };
