// @webex/react-component-activity-post-action/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Button from '@webex/react-component-button';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-activity-post-action__postActionItem__LnBvc3RB {\n  margin: 0 2px;\n  flex: 0 0 auto;\n}\n";
var styles = {"postActionItem":"react-component-activity-post-action__postActionItem__LnBvc3RB"};
styleInject(css_248z);

const propTypes = {
  iconColor: PropTypes.string,
  iconType: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  title: PropTypes.string
};
const defaultProps = {
  iconColor: 'black',
  title: ''
};

function ActivityPostAction(props) {
  return /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-post-action-item', styles.postActionItem)
  }, /*#__PURE__*/React.createElement(Button, {
    iconColor: props.iconColor,
    iconType: props.iconType,
    onClick: props.onClick,
    title: props.title
  }));
}

ActivityPostAction.propTypes = propTypes;
ActivityPostAction.defaultProps = defaultProps;

export default ActivityPostAction;
