// @webex/react-component-audio/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-audio__audio__LmF1ZGlv {\n  width: 100%;\n  height: 100%;\n}\n";
var styles = {"audio":"react-component-audio__audio__LmF1ZGlv"};
styleInject(css_248z);

const propTypes = {
  srcObject: PropTypes.object.isRequired
};

function Audio({
  srcObject
}) {
  // Need autoPlay to start the video automagically
  if (!srcObject) {
    return null;
  }

  function getEl(el) {
    const element = el;

    if (element && srcObject) {
      element.srcObject = srcObject;
    }

    return element;
  }

  return (
    /*#__PURE__*/
    // eslint-disable-reason Cannot provide captions for streaming audio yet
    // eslint-disable-next-line jsx-a11y/media-has-caption
    React.createElement("audio", {
      autoPlay: true,
      className: classNames(styles.audio),
      ref: getEl
    })
  );
}

Audio.propTypes = propTypes;

export default Audio;
