// @webex/react-component-error-display/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-error-display__errorDisplay__LmVycm9y {\n  display: flex;\n  width: 100%;\n  height: 100%;\n  background: #f5f5f6;\n}\n\n.react-component-error-display__content__LmVycm9y {\n  display: flex;\n  width: 100%;\n  text-align: center;\n  justify-content: center;\n  flex-direction: column;\n  align-items: center;\n}\n\n.react-component-error-display__transparent__LmVycm9y {\n  background: rgba(255, 255, 255, 0.9);\n}\n\n.react-component-error-display__transparent__LmVycm9y {\n  background: rgba(255, 255, 255, 0.9);\n}\n\n.react-component-error-display__contentItem__LmVycm9y {\n  padding: 2px 0;\n  font-family: CiscoSans, 'Helvetica Neue', Arial, sans-serif;\n  font-weight: 100;\n}\n\n.react-component-error-display__errorTitle__LmVycm9y {\n  font-size: 16px;\n  line-height: 20px;\n  color: #4f5051;\n}\n\n.react-component-error-display__secondaryTitle__LmVycm9y {\n  font-size: 14px;\n  line-height: 16px;\n  color: #4f5051;\n}\n\n.react-component-error-display__linkButton__LmVycm9y {\n  min-width: 50%;\n  font: inherit;\n  font-size: 14px;\n  color: #0ab7d7;\n  cursor: pointer;\n  background: none;\n  border: none;\n}\n";
var styles = {"errorDisplay":"react-component-error-display__errorDisplay__LmVycm9y","content":"react-component-error-display__content__LmVycm9y","transparent":"react-component-error-display__transparent__LmVycm9y","contentItem":"react-component-error-display__contentItem__LmVycm9y","errorTitle":"react-component-error-display__errorTitle__LmVycm9y","secondaryTitle":"react-component-error-display__secondaryTitle__LmVycm9y","linkButton":"react-component-error-display__linkButton__LmVycm9y"};
styleInject(css_248z);

const propTypes = {
  actionTitle: PropTypes.string,
  onAction: PropTypes.func,
  secondaryTitle: PropTypes.string,
  title: PropTypes.string.isRequired,
  transparent: PropTypes.bool
};
const defaultProps = {
  actionTitle: '',
  onAction: () => {},
  secondaryTitle: '',
  transparent: false
};

function ErrorDisplay({
  actionTitle,
  onAction,
  secondaryTitle,
  title,
  transparent
}) {
  function handleAction(e) {
    e.preventDefault();
    onAction();
  }

  return /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-error-display', styles.errorDisplay, {
      [styles.transparent]: transparent
    })
  }, /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-error-content', styles.content)
  }, /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-error-title', styles.errorTitle, styles.contentItem)
  }, title), secondaryTitle && /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-error-title-secondary', styles.secondaryTitle, styles.contentItem)
  }, secondaryTitle), actionTitle && onAction && /*#__PURE__*/React.createElement("button", {
    className: classNames('webex-link-button', styles.linkButton),
    onClick: handleAction,
    onKeyPress: handleAction
  }, actionTitle)));
}

ErrorDisplay.propTypes = propTypes;
ErrorDisplay.defaultProps = defaultProps;

export default ErrorDisplay;
