// @webex/react-component-activity-menu-header/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Icon, Button } from '@momentum-ui/react';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-activity-menu-header__widgetHeader__LndpZGdl {\n  display: flex;\n  flex: 0 0 auto;\n  width: 100%;\n  height: 46px;\n}\n\n.react-component-activity-menu-header__widgetMenu__LndpZGdl {\n  width: 60px;\n  padding: 10px;\n}\n\n.react-component-activity-menu-header__widgetTitle__LndpZGdl {\n  flex: 2;\n  font-size: 14px;\n  font-weight: 600;\n  line-height: 46px;\n  color: #4f5051;\n  text-align: center;\n}\n\n.react-component-activity-menu-header__menuButton__LndpZGdl {\n  width: 60px;\n  display: flex;\n  align-items: center;\n  justify-content: center;\n}\n";
var styles = {"widgetHeader":"react-component-activity-menu-header__widgetHeader__LndpZGdl","widgetMenu":"react-component-activity-menu-header__widgetMenu__LndpZGdl","widgetTitle":"react-component-activity-menu-header__widgetTitle__LndpZGdl","menuButton":"react-component-activity-menu-header__menuButton__LndpZGdl"};
styleInject(css_248z);

const propTypes = {
  activityTypes: PropTypes.array.isRequired,
  onClose: PropTypes.func.isRequired,
  onMenuClick: PropTypes.func.isRequired,
  title: PropTypes.string
};
const defaultProps = {
  title: ''
};

function ActivityMenuHeader({
  activityTypes,
  onClose,
  onMenuClick,
  title
}) {
  const displayMenu = activityTypes && activityTypes.length > 2;
  return /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-widget-header', styles.widgetHeader)
  }, onClose && /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-widget-close', styles.menuButton)
  }, /*#__PURE__*/React.createElement(Icon, {
    ariaLabel: "Close",
    name: "icon-cancel_18",
    onClick: onClose
  })), /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-widget-title', styles.widgetTitle)
  }, title), /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-widget-menu', styles.menuButton)
  }, displayMenu && onMenuClick && /*#__PURE__*/React.createElement(Button, {
    ariaLabel: "Main Menu",
    circle: true,
    onClick: onMenuClick,
    size: 32
  }, /*#__PURE__*/React.createElement(Icon, {
    color: "black",
    name: "icon-activities_16"
  }))));
}

ActivityMenuHeader.propTypes = propTypes;
ActivityMenuHeader.defaultProps = defaultProps;

export default ActivityMenuHeader;
