// @webex/react-component-file-staging-area/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { bytesToSize } from '@webex/react-component-utils';
import ChipFile from '@webex/react-component-chip-file';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-file-staging-area__fileStagingArea__LmZpbGVT {\n  position: relative;\n  display: flex;\n  overflow-y: auto;\n  background-color: hsla(240, 5%, 96%, 0.2);\n  border-bottom: 1px solid #ebebec;\n}\n\n.react-component-file-staging-area__files__LmZpbGVT {\n  display: flex;\n  max-height: 75px;\n  margin: 5px;\n  overflow-y: auto;\n  flex: 1 1 auto;\n  flex-wrap: wrap;\n}\n\n.react-component-file-staging-area__chipContainer__LmZpbGVT {\n  width: 150px;\n  height: 50px;\n  margin: 5px;\n}\n";
var styles = {"fileStagingArea":"react-component-file-staging-area__fileStagingArea__LmZpbGVT","files":"react-component-file-staging-area__files__LmZpbGVT","chipContainer":"react-component-file-staging-area__chipContainer__LmZpbGVT"};
styleInject(css_248z);

const propTypes = {
  files: PropTypes.shape({
    items: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string,
      type: PropTypes.string,
      id: PropTypes.string,
      thumbnail: PropTypes.string,
      fileSize: PropTypes.number
    }))
  }).isRequired,
  onFileRemove: PropTypes.func.isRequired
};

function FileStagingArea({
  onFileRemove,
  files
}) {
  if (!files || !Object.keys(files).length) {
    return null;
  }

  const fileChips = []; // eslint-disable-next-line react/prop-types

  files.forEach(file => {
    const chipProps = {
      name: file.name,
      size: bytesToSize(file.fileSize),
      id: file.id,
      onRemove: onFileRemove,
      thumbnail: file.thumbnail
    };
    const chip = /*#__PURE__*/React.createElement("div", {
      className: classNames('webex-chip-container', styles.chipContainer),
      key: file.id
    }, /*#__PURE__*/React.createElement(ChipFile, chipProps));
    fileChips.push(chip);
  });
  return /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-file-staging-area', styles.fileStagingArea)
  }, /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-staged-files', styles.files)
  }, fileChips));
}

FileStagingArea.propTypes = propTypes;

export default FileStagingArea;
