// @webex/react-component-chip-base/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Icon } from '@momentum-ui/react';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-chip-base__chip__LmNoaXAg {\n  position: relative;\n  display: flex;\n  width: 100%;\n  height: 100%;\n  color: #6a6b6c;\n  background-color: #fff;\n  border: 1px solid #fff;\n  border-radius: 4px;\n}\n\n.react-component-chip-base__content__LmNoaXAg {\n  flex: 1 1 auto;\n}\n\n.react-component-chip-base__action__LmNoaXAg {\n  width: 30px;\n  display: flex;\n  flex: 0 0 auto;\n  align-items: center;\n}\n\n.react-component-chip-base__action__LmNoaXAg button {\n  font-size: 16px;\n  cursor: pointer;\n}\n\n.react-component-chip-base__action__LmNoaXAg:hover button {\n  color: #ff513d;\n}\n";
var styles = {"chip":"react-component-chip-base__chip__LmNoaXAg","content":"react-component-chip-base__content__LmNoaXAg","action":"react-component-chip-base__action__LmNoaXAg"};
styleInject(css_248z);

const propTypes = {
  children: PropTypes.node.isRequired,
  id: PropTypes.string.isRequired,
  onRemove: PropTypes.func
};
const defaultProps = {
  onRemove: () => {}
};

function ChipBase(props) {
  const {
    children,
    id,
    onRemove
  } = props;

  function handleRemove() {
    onRemove(id);
  }

  return /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-chip', styles.chip)
  }, children, /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-chip-action', styles.action)
  }, /*#__PURE__*/React.createElement(Icon, {
    ariaLabel: "Delete Share",
    name: "icon-cancel_16",
    onClick: handleRemove
  })));
}

ChipBase.propTypes = propTypes;
ChipBase.defaultProps = defaultProps;

export default ChipBase;
