// @webex/react-component-activity-text/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-activity-text__activityText__LmFjdGl2 {\n  width: 100%;\n  font-size: 14px;\n  font-weight: 300;\n  letter-spacing: 0.3px;\n  color: #000;\n  word-wrap: break-word;\n  white-space: pre-wrap;\n  transition: color 0.3s, margin-left 0.3s ease 0.5s;\n}\n\n/* MARKDOWN SUPPORT */\n.react-component-activity-text__activityText__LmFjdGl2 strong {\n  font-family: CiscoSansTT Regular, Helvetica Neue, Arial;\n  font-weight: 700;\n}\n\n.react-component-activity-text__activityText__LmFjdGl2 pre {\n  padding: 10px;\n  overflow: auto;\n  background-color: #f5f5f6;\n}\n\n.react-component-activity-text__activityText__LmFjdGl2 code {\n  font-family: Consolas, Monaco, Andale Mono, Ubuntu Mono, monospace;\n  color: #000;\n  background-color: #f5f5f6;\n}\n\n.react-component-activity-text__activityText__LmFjdGl2 h1 {\n  padding-top: 1rem;\n  font-size: 2rem;\n  line-height: 2.5rem;\n}\n\n.react-component-activity-text__activityText__LmFjdGl2 h2 {\n  padding-top: 1rem;\n  font-size: 1.5rem;\n  line-height: 1.75rem;\n}\n\n.react-component-activity-text__activityText__LmFjdGl2 h3 {\n  padding-top: 1rem;\n  font-size: 1.25rem;\n  line-height: 1.1rem;\n}\n\n.react-component-activity-text__activityText__LmFjdGl2 a {\n  color: #07c1e4;\n  text-decoration: none;\n}\n\n.react-component-activity-text__activityText__LmFjdGl2 p {\n  margin: 0;\n}\n\n.react-component-activity-text__activityText__LmFjdGl2 p:not(:last-of-type) {\n  margin-bottom: 10px;\n}\n\n.react-component-activity-text__activityText__LmFjdGl2 ol, .react-component-activity-text__activityText__LmFjdGl2 pre, .react-component-activity-text__activityText__LmFjdGl2 ul {\n  margin: 1em 0!important;\n}\n\n.react-component-activity-text__activityText__LmFjdGl2 ul {\n  margin-left: 2em !important;\n}\n";
var styles = {"activityText":"react-component-activity-text__activityText__LmFjdGl2"};
styleInject(css_248z);

const propTypes = {
  content: PropTypes.string,
  displayName: PropTypes.string,
  renderedComponent: PropTypes.object
};
const defaultProps = {
  content: '',
  displayName: '',
  renderedComponent: null
};

function ActivityText({
  content,
  displayName,
  renderedComponent
}) {
  if (renderedComponent) {
    return /*#__PURE__*/React.createElement("div", {
      className: classNames('webex-activity-text', styles.activityText)
    }, renderedComponent);
  }

  if (content) {
    const htmlContent = {
      __html: content || displayName
    };
    return /*#__PURE__*/React.createElement("div", {
      className: classNames('webex-activity-text', styles.activityText) // eslint-disable-reason content is considered safe from server
      // eslint-disable-next-line react/no-danger
      ,
      dangerouslySetInnerHTML: htmlContent
    });
    /* eslint-enable react/no-danger */
  }

  return /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-activity-text', styles.activityText)
  }, displayName);
}

ActivityText.propTypes = propTypes;
ActivityText.defaultProps = defaultProps;

export default ActivityText;
