// @webex/react-component-typing-avatar/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import PresenceAvatar from '@webex/react-container-presence-avatar';
import classNames from 'classnames';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-typing-avatar__typingAvatar__LnR5cGlu {\n  position: relative;\n  width: 25px;\n  height: 25px;\n  margin: 0 2px;\n}\n\n.react-component-typing-avatar__avatar__LnR5cGlu {\n  display: flex;\n  position: absolute;\n  top: 0;\n  left: 0;\n  z-index: 2;\n  width: 25px;\n  height: 25px;\n}\n\n/* Base styles for the element that has a tooltip */\n[data-tooltip] {\n  display: inline-block;\n  position: relative;\n  cursor: pointer;\n}\n/* Tooltip styling */\n[data-tooltip]:before {\n  position: absolute;\n  bottom: 100%;\n  left: 50%;\n  transform: translateX(-50%);\n  width: 150px;\n  background-color: #000;\n  font-size: 12px;\n  content: attr(data-tooltip);\n  opacity: 0;\n  color: #fff;\n  padding: 8px 12px;\n  margin-bottom: 12px;\n  text-align: center;\n  border-radius: 4px;\n  white-space: pre-wrap;\n}\n\n/* Tooltip Bottom Arrow */\n[data-tooltip]:after {\n  content: '';\n  opacity: 0;\n  position: absolute;\n  width: 0;\n  height: 0;\n  border-color: transparent;\n  border-style: solid;\n  left: 50%;\n  margin-left: -6px;\n  bottom: 31px;\n  border-width: 6px 6px 0;\n  border-top-color: #000;\n}\n\n/* Show the tooltip when hovering */\n[data-tooltip]:hover:before,\n[data-tooltip]:hover:after {\n  opacity: 1;\n}\n";
var styles = {"typingAvatar":"react-component-typing-avatar__typingAvatar__LnR5cGlu","avatar":"react-component-typing-avatar__avatar__LnR5cGlu"};
styleInject(css_248z);

const propTypes = {
  avatarId: PropTypes.string,
  isTyping: PropTypes.bool,
  name: PropTypes.string.isRequired
};
const defaultProps = {
  avatarId: '',
  isTyping: false
};

function TypingAvatar({
  avatarId,
  isTyping,
  name
}) {
  return /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-typing-avatar', styles.typingAvatar),
    "data-tooltip": name
  }, /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-avatar', styles.avatar)
  }, /*#__PURE__*/React.createElement(PresenceAvatar, {
    avatarId: avatarId,
    isTyping: isTyping,
    name: name,
    size: 24
  })));
}

TypingAvatar.propTypes = propTypes;
TypingAvatar.defaultProps = defaultProps;

export default TypingAvatar;
