// @webex/react-component-chip-file/es/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import ChipBase from '@webex/react-component-chip-base';
import { Icon } from '@momentum-ui/react';

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".react-component-chip-file__info__LmluZm8s,\n.react-component-chip-file__name__LmluZm8s,\n.react-component-chip-file__meta__LmluZm8s {\n  width: 100%;\n  overflow: hidden;\n  font-size: 10px;\n  text-overflow: ellipsis;\n  white-space: nowrap;\n}\n\n.react-component-chip-file__info__LmluZm8s {\n  display: flex;\n  padding: 0 5px;\n  margin: auto 0;\n  flex-direction: column;\n  justify-content: center;\n}\n\n.react-component-chip-file__name__LmluZm8s {\n  color: #666;\n}\n\n.react-component-chip-file__meta__LmluZm8s {\n  color: #c2c2c2;\n}\n\n.react-component-chip-file__type__LmluZm8s,\n.react-component-chip-file__size__LmluZm8s {\n  display: inline-block;\n  margin-right: 10px;\n}\n\n.react-component-chip-file__icon__LmluZm8s {\n  height: 36px;\n}\n\n.react-component-chip-file__thumbnail__LmluZm8s {\n  display: flex;\n  max-width: 50px;\n  height: 50px;\n  flex: 0 0 auto;\n  align-items: center;\n}\n\n.react-component-chip-file__thumbnail__LmluZm8s img {\n  max-width: 100%;\n  max-height: 100%;\n}\n";
var styles = {"info":"react-component-chip-file__info__LmluZm8s","name":"react-component-chip-file__name__LmluZm8s","meta":"react-component-chip-file__meta__LmluZm8s","type":"react-component-chip-file__type__LmluZm8s","size":"react-component-chip-file__size__LmluZm8s","icon":"react-component-chip-file__icon__LmluZm8s","thumbnail":"react-component-chip-file__thumbnail__LmluZm8s"};
styleInject(css_248z);

const propTypes = {
  name: PropTypes.string.isRequired,
  size: PropTypes.string,
  thumbnail: PropTypes.string
};
const defaultProps = {
  size: '',
  thumbnail: ''
};

function ChipFile(props) {
  const {
    name,
    size,
    thumbnail
  } = props;
  let icon;

  if (!thumbnail) {
    icon = /*#__PURE__*/React.createElement("div", {
      className: classNames('webex-file-icon', styles.icon)
    }, /*#__PURE__*/React.createElement(Icon, {
      name: "icon-document_36"
    }));
  }

  return /*#__PURE__*/React.createElement(ChipBase, props, /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-file-thumbnail', styles.thumbnail)
  }, icon, /*#__PURE__*/React.createElement("img", {
    alt: "",
    src: thumbnail
  })), /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-file-info', styles.info)
  }, /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-file-name', styles.name)
  }, name), /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-file-meta', styles.meta)
  }, /*#__PURE__*/React.createElement("div", {
    className: classNames('webex-file-size', styles.size)
  }, size))));
}

ChipFile.propTypes = propTypes;
ChipFile.defaultProps = defaultProps;

export default ChipFile;
